File: Array/zhangjunwei/queue/ValidParentheses_6_16.c



#include <stdbool.h>
#include <string.h>
/**
 * 在 ASCII中
 * （  == 40
 *  ） == 41
 *  [ == 91
 *  ] == 93
 *  { == 123
 *  } == 125
 */
bool isValid(char *s) {
    int len = strlen(s);
    if(len==0)return true;
    //() [] {} 必须是成对出现的  len 必须为偶数
    if (len&1==1)return false; //len%2==1
    char stack[len];
    int top = -1;//表示栈空
    for (int i=0;i<len;i++){
        if (s[i]=='(' || s[i]=='['||s[i]=='{'){//是左括号时 入栈
            stack[++top] = s[i];
        } else if (top==-1)// 不是左括号们，栈空则无法配对，不符合
            return false;
        else if (s[i] == stack[top]+1 || s[i]==stack[top]+2){//s[i]是栈顶元素的右括号们
            stack[top--] = 0;
        }
    }
    return top==-1;
}

File: Array/zhangjunwei/queue/ValidParentheses_6_16.kt
package mykotlin


class Solution1 {
    /**
     * '('，')'，'{'，'}'，'['，']'
     */
    fun isValid(s: String): Boolean {
        var charArray = s.toCharArray()
        var top = -1
        var map: Map<Char, Char> = mapOf(')' to '(', '}' to '{', ']' to '[')

        for (it in charArray) {
            if (top<0){//初始化
                top=0
                charArray[top] = it

            }else{
                if (map.get(it) == charArray[top]){//出栈
                     top--
                }else{//入栈
                    charArray[++top] = it
                }
            }
        }



        return top == -1
    }

}

fun main() {
    val res = Solution1().isValid("()[]{}")
    println(res)
}